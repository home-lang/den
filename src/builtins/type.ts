import type { BuiltinCommand, CommandResult, Shell } from './types'
import { access } from 'node:fs/promises'
import { join } from 'node:path'

/**
 * Type command - show the type of a command
 * Identifies if a command is an alias, builtin, file, or not found
 */
export const typeCommand: BuiltinCommand = {
  name: 'type',
  description: 'Display the type of a command',
  usage: 'type [-afptP] [name ...]',
  async execute(args: string[], shell: Shell): Promise<CommandResult> {
    const start = performance.now()

    if (args.length === 0) {
      return {
        exitCode: 1,
        stdout: '',
        stderr: 'type: missing argument\n',
        duration: performance.now() - start,
      }
    }

    const results: string[] = []
    let allFound = true
    let showAll = false
    let fileOnly = false
    let noPath = false
    let showPath = false

    // Parse options
    while (args[0]?.startsWith('-')) {
      const arg = args.shift()!
      if (arg === '--')
        break

      for (let i = 1; i < arg.length; i++) {
        const flag = arg[i]
        switch (flag) {
          case 'a': showAll = true; break
          case 'f': fileOnly = true; break
          case 'p': noPath = true; break
          case 'P': showPath = true; break
          case 't': fileOnly = true; noPath = true; break
          default:
            return {
              exitCode: 1,
              stdout: '',
              stderr: `type: -${flag}: invalid option\ntype: usage: type [-afptP] name [name ...]\n`,
              duration: performance.now() - start,
            }
        }
      }
    }

    if (args.length === 0) {
      return {
        exitCode: 1,
        stdout: '',
        stderr: 'type: missing argument\n',
        duration: performance.now() - start,
      }
    }

    for (const name of args) {
      if (!name)
        continue

      let found = false

      // Check for alias
      if (!fileOnly && shell.aliases[name]) {
        found = true
        if (noPath) {
          results.push('alias')
        }
        else {
          results.push(`${name} is aliased to '${shell.aliases[name]}'`)
        }
        if (!showAll)
          continue
      }

      // Check for builtin
      if (!fileOnly && shell.builtins.has(name)) {
        found = true
        if (noPath) {
          results.push('builtin')
        }
        else {
          results.push(`${name} is a shell builtin`)
        }
        if (!showAll)
          continue
      }

      // Check for file in PATH
      const pathDirs = (shell.environment.PATH || '').split(':')
      let filePath = ''

      if (name.includes('/')) {
        // Absolute or relative path provided
        try {
          await access(name)
          filePath = name
          found = true
        }
        catch {
          // File not found, continue to next check
        }
      }
      else {
        // Search in PATH
        for (const dir of pathDirs) {
          if (!dir)
            continue
          const fullPath = join(dir, name)
          try {
            await access(fullPath)
            filePath = fullPath
            found = true
            break
          }
          catch {
            // File not found in this directory, continue searching
          }
        }
      }

      if (filePath) {
        if (noPath) {
          results.push('file')
        }
        else if (showPath) {
          results.push(filePath)
        }
        else {
          results.push(`${name} is ${filePath}`)
        }
        continue
      }

      // If we get here, command was not found
      if (!found) {
        allFound = false
        results.push(`type: ${name}: not found`)
      }
    }

    return {
      exitCode: allFound ? 0 : 1,
      stdout: `${results.join('\n')}\n`,
      stderr: '',
      duration: performance.now() - start,
    }
  },
}
