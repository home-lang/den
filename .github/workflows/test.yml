name: Tests

on:
  push:
    branches: [main, develop]
  pull_request:
    branches: [main, develop]

jobs:
  test:
    name: Run Tests
    runs-on: ${{ matrix.os }}
    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, macos-latest]
        zig-version: ['0.15.1']

    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Setup Zig
        uses: goto-bus-stop/setup-zig@v2
        with:
          version: ${{ matrix.zig-version }}

      - name: Cache Zig artifacts
        uses: actions/cache@v3
        with:
          path: |
            ~/.cache/zig
            .zig-cache
          key: ${{ runner.os }}-zig-${{ matrix.zig-version }}-${{ hashFiles('build.zig') }}
          restore-keys: |
            ${{ runner.os }}-zig-${{ matrix.zig-version }}-

      - name: Build
        run: zig build

      - name: Run tokenizer tests
        run: zig build test-tokenizer

      - name: Run parser tests
        run: zig build test-parser

      - name: Run expander tests
        run: zig build test-expander

      - name: Run executor tests
        run: zig build test-executor

      - name: Run plugin tests
        run: zig build test-all-plugins

      - name: Run hook tests
        run: zig build test-all-hooks

      - name: Run theme tests
        run: zig build test-theme

      - name: Upload test results
        if: always()
        uses: actions/upload-artifact@v3
        with:
          name: test-results-${{ matrix.os }}
          path: |
            zig-out/test/
            .zig-cache/

  coverage:
    name: Code Coverage
    runs-on: ubuntu-latest
    needs: test

    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Setup Zig
        uses: goto-bus-stop/setup-zig@v2
        with:
          version: '0.15.1'

      - name: Generate coverage
        run: |
          zig build test -Dtest-coverage=true || true

      - name: Upload coverage to Codecov
        uses: codecov/codecov-action@v3
        with:
          files: ./coverage.txt
          flags: unittests
          name: codecov-umbrella

  lint:
    name: Lint and Format Check
    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Setup Zig
        uses: goto-bus-stop/setup-zig@v2
        with:
          version: '0.15.1'

      - name: Check formatting
        run: zig fmt --check .

  build-release:
    name: Build Release
    runs-on: ${{ matrix.os }}
    needs: [test, lint]
    strategy:
      matrix:
        os: [ubuntu-latest, macos-latest]

    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Setup Zig
        uses: goto-bus-stop/setup-zig@v2
        with:
          version: '0.15.1'

      - name: Build release
        run: zig build -Doptimize=ReleaseSafe

      - name: Upload artifacts
        uses: actions/upload-artifact@v3
        with:
          name: den-${{ matrix.os }}
          path: zig-out/bin/den
