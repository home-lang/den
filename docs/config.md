# Configuration

Krusty is configured via a `krusty.config.ts` (or `.js`) file. This page highlights key options, with emphasis on system modules now driven by centralized configuration.

## Quick start

```ts
import type { KrustyConfig } from 'krusty/src/types'

const config: KrustyConfig = {
  prompt: { format: '{user}@{host} {path}{git} {modules} {symbol} ' },
  modules: {
    os: {
      enabled: false,
      format: 'on {symbol} {name}',
      symbol: '💻',
      symbols: { darwin: '', linux: '🐧', win32: '🪟' },
    },
    hostname: { enabled: true, format: '@{hostname}', ssh_only: true, showOnLocal: false },
    username: { enabled: true, format: '{username}', show_always: false, showOnLocal: false, root_format: '{username}' },
    directory: { enabled: true, format: '{path}', truncation_length: 3, truncate_to_repo: true, home_symbol: '~', readonly_symbol: '🔒' },
    battery: {
      enabled: true,
      format: '{symbol} {percentage}%',
      // Legacy keys still supported
      full_symbol: '🔋',
      charging_symbol: '🔌',
      discharging_symbol: '🔋',
      unknown_symbol: '🔋',
      empty_symbol: '🪫',
      // New unified keys
      symbol: '🔋',
      symbol_charging: '🔌',
      symbol_low: '🪫',
    },
    cmd_duration: { enabled: true, format: 'took {duration}', min_time: 2000, min_ms: 2000, show_milliseconds: false },
    time: { enabled: false, format: '{symbol} {time}', symbol: '🕐', locale: undefined, options: { hour: '2-digit', minute: '2-digit' } },
    memory_usage: { enabled: false, format: '🐏 {ram}', threshold: 75, symbol: '🐏' },
    nix_shell: { enabled: true, format: 'via {symbol} {state}', symbol: '❄️' },
  },
}

export default config
```

## Placeholders

Most module `format` strings accept placeholders. Common ones:

- `os`: {symbol}, {name}
- `hostname`: {hostname}
- `username`: {username}
- `directory`: {path}
- `battery`: {symbol}, {percentage}
- `cmd_duration`: {duration}
- `time`: {symbol}, {time}
- `nix_shell`: {symbol}, {state}

## Backward compatibility

Legacy keys like `battery.full_symbol`, `cmd_duration.min_time`, and `time: '🕐 {time}'` remain supported. The new unified keys and formats provide a single source of truth while preserving existing configs.

## Tips

- Symbols and colors are theme-driven; modules avoid hardcoded colors.
- You can override per-platform OS symbols via `os.symbols`.
- Set `hostname.showOnLocal`/`username.showOnLocal` to show/hide on local sessions.
