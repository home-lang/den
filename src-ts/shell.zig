const std = @import("std");

pub const Shell = struct {
    allocator: std.mem.Allocator,
    running: bool,

    pub fn init(allocator: std.mem.Allocator) !Shell {
        return Shell{
            .allocator = allocator,
            .running = false,
        };
    }

    pub fn deinit(self: *Shell) void {
        _ = self;
    }

    pub fn run(self: *Shell) !void {
        self.running = true;
        const stdout = std.io.getStdOut().writer();

        try stdout.print("Den shell initialized!\n", .{});
        try stdout.print("Type 'exit' to quit.\n\n", .{});

        while (self.running) {
            try stdout.print("den> ", .{});

            // Read input
            const stdin = std.io.getStdIn().reader();
            var buffer: [1024]u8 = undefined;

            if (try stdin.readUntilDelimiterOrEof(&buffer, '\n')) |line| {
                const trimmed = std.mem.trim(u8, line, &std.ascii.whitespace);

                if (trimmed.len == 0) continue;

                if (std.mem.eql(u8, trimmed, "exit")) {
                    self.running = false;
                    try stdout.print("Goodbye!\n", .{});
                    break;
                }

                try stdout.print("Command: {s}\n", .{trimmed});
            } else {
                // EOF (Ctrl+D)
                self.running = false;
                try stdout.print("\nGoodbye!\n", .{});
            }
        }
    }
};

test "shell initialization" {
    const allocator = std.testing.allocator;
    var sh = try Shell.init(allocator);
    defer sh.deinit();

    try std.testing.expect(!sh.running);
}
