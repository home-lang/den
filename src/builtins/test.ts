import type { BuiltinCommand, CommandResult, Shell } from './types'
import { access, stat } from 'node:fs/promises'

/**
 * Test command - evaluate conditional expressions
 * Implements the POSIX test command functionality
 */
export const testCommand: BuiltinCommand = {
  name: 'test',
  description: 'Evaluate conditional expressions',
  usage: 'test [expression]',
  async execute(args: string[], shell: Shell): Promise<CommandResult> {
    const start = performance.now()

    // Handle empty arguments
    if (args.length === 0) {
      return {
        exitCode: 1,
        stdout: '',
        stderr: '',
        duration: performance.now() - start,
      }
    }

    // Handle the [ ] syntax (last argument should be ']')
    if (args[0] === '[') {
      if (args[args.length - 1] !== ']') {
        return {
          exitCode: 2,
          stdout: '',
          stderr: 'test: missing `]\'\n',
          duration: performance.now() - start,
        }
      }
      args = args.slice(1, -1) // Remove [ and ]
    }

    // Helper function to evaluate a test expression
    const evaluateTest = async (tokens: string[]): Promise<{ result: boolean, consumed: number }> => {
      if (tokens.length === 0) {
        return { result: false, consumed: 0 }
      }

      let pos = 0

      // Helper to get next token
      const next = () => tokens[pos++]
      const peek = () => tokens[pos]
      const eof = () => pos >= tokens.length

      // Parse a primary expression
      const parsePrimary = async (): Promise<boolean> => {
        const token = next()

        // Handle unary operators
        if (token === '!') {
          return !(await parsePrimary())
        }

        if (token === '(') {
          const result = await parseOr()
          if (next() !== ')') {
            throw new Error('syntax error: missing `)\'\n')
          }
          return result
        }

        // Handle string tests
        if (token === '-n') {
          return next().length > 0
        }

        if (token === '-z') {
          return next().length === 0
        }

        // Handle file tests
        if (token.startsWith('-')) {
          const arg = next()
          if (arg === undefined) {
            throw new Error(`test: ${token}: argument expected\n`)
          }

          try {
            const stats = await stat(arg)

            switch (token) {
              case '-b': return stats.isBlockDevice()
              case '-c': return stats.isCharacterDevice()
              case '-d': return stats.isDirectory()
              case '-e': return true
              case '-f': return stats.isFile()
              case '-g': return (stats.mode & 0o2000) !== 0 // setgid bit
              case '-G': return stats.gid === process.getgid?.()
              case '-h':
              case '-L': return stats.isSymbolicLink()
              case '-k': return (stats.mode & 0o100) !== 0 // sticky bit
              case '-O': return stats.uid === process.getuid?.()
              case '-p': return stats.isFIFO()
              case '-r': {
                try {
                  await access(arg, 0o400) // Readable by user
                  return true
                }
                catch {
                  return false
                }
              }
              case '-s': return stats.size > 0
              case '-S': return stats.isSocket()
              case '-t': return process.stdin.isTTY
              case '-u': return (stats.mode & 0o4000) !== 0 // setuid bit
              case '-w': {
                try {
                  await access(arg, 0o200) // Writable by user
                  return true
                }
                catch {
                  return false
                }
              }
              case '-x': {
                try {
                  await access(arg, 0o100) // Executable by user
                  return true
                }
                catch {
                  return false
                }
              }
              default:
                throw new Error(`test: ${token}: unary operator expected\n`)
            }
          }
          catch (error) {
            if (error.code === 'ENOENT') {
              return false
            }
            throw error
          }
        }

        // Handle string comparisons
        const nextToken = peek()

        if (nextToken === '=') {
          next() // consume '='
          return token === next()
        }

        if (nextToken === '!=') {
          next() // consume '!='
          return token !== next()
        }

        // Handle numeric comparisons
        if (nextToken === '-eq') {
          next() // consume '-eq'
          return Number(token) === Number(next())
        }

        if (nextToken === '-ne') {
          next() // consume '-ne'
          return Number(token) !== Number(next())
        }

        if (nextToken === '-lt') {
          next() // consume '-lt'
          return Number(token) < Number(next())
        }

        if (nextToken === '-le') {
          next() // consume '-le'
          return Number(token) <= Number(next())
        }

        if (nextToken === '-gt') {
          next() // consume '-gt'
          return Number(token) > Number(next())
        }

        if (nextToken === '-ge') {
          next() // consume '-ge'
          return Number(token) >= Number(next())
        }

        // If we get here, it's just a string
        return token.length > 0
      }

      // Parse AND expressions
      const parseAnd = async (): Promise<boolean> => {
        let result = await parsePrimary()

        while (!eof() && peek() === '-a') {
          next() // consume '-a'
          result = result && (await parsePrimary())
        }

        return result
      }

      // Parse OR expressions
      const parseOr = async (): Promise<boolean> => {
        let result = await parseAnd()

        while (!eof() && peek() === '-o') {
          next() // consume '-o'
          result = result || (await parseAnd())
        }

        return result
      }

      const result = await parseOr()
      return { result, consumed: pos }
    }

    try {
      const { result, consumed } = await evaluateTest(args)

      // Check if we consumed all tokens
      if (consumed < args.length) {
        return {
          exitCode: 2,
          stdout: '',
          stderr: `test: too many arguments\n`,
          duration: performance.now() - start,
        }
      }

      return {
        exitCode: result ? 0 : 1,
        stdout: '',
        stderr: '',
        duration: performance.now() - start,
      }
    }
    catch (error) {
      return {
        exitCode: 2,
        stdout: '',
        stderr: `test: ${(error as Error).message}`,
        duration: performance.now() - start,
      }
    }
  },
}
