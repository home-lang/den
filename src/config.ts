import type { KrustyConfig } from './types'
import { loadConfig } from 'bunfig'

export const defaultConfig: KrustyConfig = {
  verbose: false,
  streamOutput: true,
  logging: {
    prefixes: {
      debug: 'DEBUG',
      info: 'INFO',
      warn: 'WARN',
      error: 'ERROR',
    },
  },
  prompt: {
    format: '{user}@{host} {path}{git} {symbol} ',
    showGit: true,
    showTime: false,
    showUser: true,
    showHost: true,
    showPath: true,
    showExitCode: true,
    transient: false,
  },
  history: {
    maxEntries: 10000,
    file: '~/.krusty_history',
    ignoreDuplicates: true,
    ignoreSpace: true,
    searchMode: 'fuzzy',
  },
  completion: {
    // Enable/disable completion system
    enabled: true,
    // Case sensitivity for completions
    caseSensitive: false,
    // Show descriptions with completions
    showDescriptions: true,
    // Maximum number of suggestions to show
    maxSuggestions: 10,
    // Enable completion caching
    cache: {
      enabled: true,
      // Cache expiration in milliseconds (1 hour)
      ttl: 60 * 60 * 1000,
      // Maximum number of completions to cache
      maxEntries: 1000,
    },
    // Context-aware completion settings
    context: {
      // Enable context-aware completions
      enabled: true,
      // Maximum depth to analyze context
      maxDepth: 3,
      // File types to analyze for context
      fileTypes: ['.ts', '.js', '.tsx', '.jsx', '.json', '.md'],
    },
    // Command-specific completion settings
    commands: {
      git: {
        // Enable git completions
        enabled: true,
        // Include porcelain commands
        includePorcelain: true,
        // Include plumbing commands
        includePlumbing: false,
      },
      npm: {
        // Enable npm completions
        enabled: true,
        // Include script names from package.json
        includeScripts: true,
      },
    },
  },
  aliases: {
    // Stage all changes and open a commit message prompt
    commit: 'git add .; git commit -m',
    // Work-in-progress: create a WIP commit and push with styled, informative output
    // - Show concise status and staged diff before committing
    // - Suppress commit's own summary (-q) to avoid duplicate info
    // - Skip commit/push if no staged changes are present
    // - Only run post-commit log and push if commit succeeds
    wip: 'printf \'\\x1b[2m\\x1b[36m─── WIP start ───\\x1b[0m\\n\'; git --no-pager -c color.ui=always status -sb; git -c color.ui=always add -A; printf \'\\x1b[2mstaged summary\\x1b[0m\\n\'; git --no-pager -c color.ui=always diff --cached --stat; git diff --cached --quiet && printf \'\\x1b[2m\\x1b[33mno changes to commit; skipping push\\x1b[0m\\n\' || git -c color.ui=always commit -m \'chore: wip\' -q && printf \'\\x1b[2mcommit (last)\\x1b[0m\\n\' && git --no-pager -c color.ui=always log -1 --oneline && printf \'\\x1b[2m\\x1b[36m─── pushing ───\\x1b[0m\\n\' && git -c color.ui=always push; printf \'\\x1b[2m\\x1b[32m─── done ───\\x1b[0m\\n\'',
    // Push current branch
    push: 'git push',
  },
  environment: {},
  // Plugin system configuration
  plugins: [
    // Example plugin configuration:
    // {
    //   name: 'example-plugin',
    //   enabled: true,
    //   path: './plugins/example',
    //   config: {}
    // }
  ],
  theme: {
    // Theme name to use (must match a theme in themes/ directory)
    name: 'default',
    // Auto-detect system color scheme
    autoDetectColorScheme: true,
    // Default color scheme (light/dark/auto)
    defaultColorScheme: 'auto',
    // Enable right prompt
    enableRightPrompt: true,
    // Git status in prompt settings
    gitStatus: {
      enabled: true,
      showStaged: true,
      showUnstaged: true,
      showUntracked: true,
      showAheadBehind: true,
      format: '({branch}{ahead}{behind}{staged}{unstaged}{untracked})',
    },
    // Prompt configuration
    prompt: {
      left: '\u001B[32m{user}@{host}\u001B[0m \u001B[34m{path}\u001B[0m {git}{symbol} ',
      right: '{time}{jobs}{status}',
      continuation: '... ',
      error: '\u001B[31m{code}\u001B[0m',
    },
    // Colors configuration
    colors: {
      primary: '#00D9FF',
      secondary: '#FF6B9D',
      success: '#00FF88',
      warning: '#FFD700',
      error: '#FF4757',
      info: '#74B9FF',
      // Git status colors
      git: {
        branch: '#F8F8F2',
        ahead: '#50FA7B',
        behind: '#FF5555',
        staged: '#50FA7B',
        unstaged: '#FFB86C',
        untracked: '#FF79C6',
        conflict: '#FF5555',
      },
    },
    // Font configuration
    font: {
      // Default to system monospace font stack with fallbacks
      family: 'ui-monospace, SFMono-Regular, SF Mono, Menlo, Consolas, Liberation Mono, monospace',
      size: 14,
      weight: 'normal',
      lineHeight: 1.4,
      ligatures: false,
    },
    // Symbol configuration
    symbols: {
      prompt: '❯',
      continuation: '…',
      git: {
        branch: '',
        ahead: '⇡',
        behind: '⇣',
        staged: '●',
        unstaged: '○',
        untracked: '?',
        conflict: '✗',
      },
    },
  },
  modules: {
    // Language modules - enabled by default
    bun: { enabled: true, format: 'via {symbol} {version}', symbol: '🥟' },
    deno: { enabled: true, format: 'via {symbol} {version}', symbol: '🦕' },
    nodejs: { enabled: true, format: 'via {symbol} {version}', symbol: '⬢' },
    python: { enabled: true, format: 'via {symbol} {version}', symbol: '🐍' },
    golang: { enabled: true, format: 'via {symbol} {version}', symbol: '🐹' },
    java: { enabled: true, format: 'via {symbol} {version}', symbol: '☕' },
    kotlin: { enabled: true, format: 'via {symbol} {version}', symbol: '🅺' },
    php: { enabled: true, format: 'via {symbol} {version}', symbol: '🐘' },
    ruby: { enabled: true, format: 'via {symbol} {version}', symbol: '💎' },
    swift: { enabled: true, format: 'via {symbol} {version}', symbol: '🐦' },
    zig: { enabled: true, format: 'via {symbol} {version}', symbol: '⚡' },
    lua: { enabled: true, format: 'via {symbol} {version}', symbol: '🌙' },
    perl: { enabled: true, format: 'via {symbol} {version}', symbol: '🐪' },
    rlang: { enabled: true, format: 'via {symbol} {version}', symbol: '📊' },
    dotnet: { enabled: true, format: 'via {symbol} {version}', symbol: '.NET' },
    erlang: { enabled: true, format: 'via {symbol} {version}', symbol: 'E' },
    c: { enabled: true, format: 'via {symbol} {version}', symbol: 'C' },
    cpp: { enabled: true, format: 'via {symbol} {version}', symbol: 'C++' },
    cmake: { enabled: true, format: 'via {symbol} {version}', symbol: '△' },
    terraform: { enabled: true, format: 'via {symbol} {version}', symbol: '💠' },
    pulumi: { enabled: true, format: 'via {symbol} {version}', symbol: '🧊' },

    // Cloud modules
    aws: { enabled: true, format: 'on {symbol} {profile}({region})', symbol: '☁️' },
    azure: { enabled: true, format: 'on {symbol} {subscription}', symbol: '󰠅' },
    gcloud: { enabled: true, format: 'on {symbol} {project}', symbol: '☁️' },

    // Git modules
    git_branch: {
      enabled: true,
      format: 'on {symbol} {branch}',
      symbol: '',
      truncation_length: 20,
      truncation_symbol: '…',
    },
    git_commit: { enabled: true, format: '({hash})', commit_hash_length: 7 },
    git_state: { enabled: true, format: '({state})' },
    git_status: { enabled: true, format: '[{status}]' },
    git_metrics: { enabled: true, format: '({metrics})' },

    // System modules
    os: { enabled: false, format: 'on {symbol} {name}' }, // Disabled by default
    hostname: { enabled: true, format: '@{hostname}', ssh_only: true },
    directory: {
      enabled: true,
      format: '{path}',
      truncation_length: 3,
      truncate_to_repo: true,
      home_symbol: '~',
    },
    username: { enabled: true, format: '{username}', show_always: false },
    shell: { enabled: false, format: '{indicator}' }, // Disabled by default
    battery: {
      enabled: true,
      format: '{symbol} {percentage}%',
      full_symbol: '🔋',
      charging_symbol: '🔌',
      discharging_symbol: '🔋',
      unknown_symbol: '🔋',
      empty_symbol: '🪫',
    },
    cmd_duration: {
      enabled: true,
      format: 'took {duration}',
      min_time: 2000,
      show_milliseconds: false,
    },
    memory_usage: {
      enabled: false, // Disabled by default
      format: '🐏 {ram}',
      threshold: 75,
      symbol: '🐏',
    },
    time: { enabled: false, format: '🕐 {time}' }, // Disabled by default
    nix_shell: {
      enabled: true,
      format: 'via {symbol} {state}',
      symbol: '❄️',
      impure_msg: 'impure',
      pure_msg: 'pure',
      unknown_msg: 'shell',
    },

    // Custom modules can be added by users in their config
    env_var: {},
    custom: {},
  },

  // Hooks configuration
  hooks: {
    // Shell lifecycle hooks
    'shell:init': [],
    'shell:start': [],
    'shell:stop': [],
    'shell:exit': [],

    // Command hooks
    'command:before': [],
    'command:after': [],
    'command:error': [],

    // Prompt hooks
    'prompt:before': [],
    'prompt:after': [],
    'prompt:render': [],

    // Directory hooks
    'directory:change': [],
    'directory:enter': [],
    'directory:leave': [],

    // History hooks
    'history:add': [],
    'history:search': [],

    // Completion hooks
    'completion:before': [],
    'completion:after': [],
  },
}

// Use a function to avoid top-level await issues
export const config: KrustyConfig = (() => {
  try {
    return loadConfig({
      name: 'krusty',
      defaultConfig,
    }) as any
  }
  catch {
    return defaultConfig
  }
})()

// Provide a reusable loader that always fetches the latest config from disk
export async function loadKrustyConfig(): Promise<KrustyConfig> {
  return await loadConfig({
    name: 'krusty',
    defaultConfig,
  })
}
