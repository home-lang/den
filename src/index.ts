export * from './builtins'
export * from './completion'
export * from './config'
export * from './history'
export * from './hooks'
export * from './modules'
export * from './parser'
export * from './plugins/plugin-manager'
export * from './prompt'
export * from './shell'
export * from './types'
