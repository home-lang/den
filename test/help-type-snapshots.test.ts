import type { KrustyConfig } from '../src/types'
import { afterEach, beforeEach, describe, expect, it } from 'bun:test'
import { defaultConfig } from '../src/config'
import { KrustyShell } from '../src/shell'

// Snapshot tests for help and type outputs

describe('snapshot: help and type', () => {
  let shell: KrustyShell
  let cfg: KrustyConfig

  beforeEach(() => {
    cfg = { ...defaultConfig, verbose: false, history: { ...defaultConfig.history, file: `/tmp/test_history_snap_${Math.random().toString(36).slice(2)}` } }
    shell = new KrustyShell(cfg)
  })

  afterEach(() => shell.stop())

  it('help (no args) output', async () => {
    const res = await shell.execute('help')
    expect(res.exitCode).toBe(0)
    // Normalize dynamic whitespace
    const normalized = res.stdout.replace(/\s+$/gm, '')
    expect(normalized).toMatchSnapshot()
  })

  it('help printf output', async () => {
    const res = await shell.execute('help printf')
    expect(res.exitCode).toBe(0)
    const normalized = res.stdout.replace(/\s+$/gm, '')
    expect(normalized).toMatchSnapshot()
  })

  it('type builtins and external', async () => {
    const r1 = await shell.execute('type printf')
    const r2 = await shell.execute('type env')
    // "echo" is builtin too
    const r3 = await shell.execute('type echo')
    expect([r1.exitCode, r2.exitCode, r3.exitCode]).toEqual([0, 0, 0])
    const snapshot = [r1.stdout.trim(), r2.stdout.trim(), r3.stdout.trim()].join('\n---\n')
    expect(snapshot).toMatchSnapshot()
  })
})
