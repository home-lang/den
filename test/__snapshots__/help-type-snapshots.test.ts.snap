// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`snapshot: help and type help (no args) output 1`] = `
"Built-in commands:
alias        Define or display aliases
b            Run build via bun run build
bb           Run build script via bun run build
bd           Run dev via bun run dev
bf           Format code using pickier (or prettier)
bg           Resume suspended jobs in the background
bi           Install dependencies via bun install
bl           Lint code using pickier (or eslint)
bookmark     Manage directory bookmarks and navigate quickly
bm           Manage directory bookmarks and navigate quickly
br           Run script via bun run (default: start)
mark         Manage directory bookmarks and navigate quickly
jump         Manage directory bookmarks and navigate quickly
builtin      Run a shell builtin explicitly
c            Clear the screen
cd           Change the current directory
command      Run a command ignoring functions and aliases
code         Open the current directory in Visual Studio Code
copyssh      Copy ~/.ssh/id_ed25519.pub to clipboard when available, else print
dirs         Display the directory stack
disown       Remove jobs from the job table
dotfiles     Open $DOTFILES in your preferred editor ($EDITOR) or default to VS Code
echo         Display text
emptytrash   Empty the user Trash on macOS (no sudo). Fails gracefully elsewhere.
eval         Concatenate arguments and evaluate as a command
env          Print or run a command in a modified environment
exec         Execute a command
exit         Exit the shell
export       Set environment variables
false        Return unsuccessful exit status
fg           Bring a background job to the foreground
ft           Fix/Unstick macOS Touch Bar when it freezes
getopts      Parse positional parameters as options
hash         Remember or display command locations
help         Display help information
hide         Hide hidden files in Finder (macOS)
history      Display or manipulate the command history
ip           Show public IP address via OpenDNS diagnostic service
jobs         List background jobs
kill         Send a signal to a process or job
library      cd to $HOME/Library
localip      Show local IP addresses (IPv4/IPv6) from ifconfig output
pstorm       Open the current directory in PhpStorm
popd         Pop directory from stack and change to it
printf       Format and print data
pushd        Save current directory on stack and change to DIR
pwd          Print the current working directory
read         Read a line from standard input
reload       Reload configuration, hooks, and plugins (same as sourcing shell config)
reloaddns    Flush DNS cache on macOS
reload       Reload configuration, hooks, and plugins (same as sourcing shell config)
set          Set shell options and positional parameters or display variables
show         Show hidden files in Finder (macOS)
shrug        Copy ¯\\_(ツ)_/¯ to clipboard (when available) or print it
source       Execute commands from a file in the current shell environment
source       Execute commands from a file in the current shell environment
test         Evaluate conditional expressions
[            Evaluate conditional expressions (alias for test)
time         Measure command execution time
timeout      Run a command with a time limit
times        Print accumulated user and system times
trap         Trap signals and other events
true         Return successful exit status
type         Display the type of a command
umask        Get or set the file mode creation mask
unalias      Remove aliases
unset        Unset (remove) shell variables or functions
wait         Wait for background jobs or PIDs to finish
web          cd to $HOME/Code
which        Show the full path of commands
wip          Create a work-in-progress commit and optionally push it
return       Return from a function or script
break        Break out of loops
continue     Continue to next iteration of loop
local        Create local variables in functions
readonly     Mark variables as read-only
declare      Declare variables and give them attributes
Use 'help <command>' for detailed information."
`;

exports[`snapshot: help and type help printf output 1`] = `
"printf: Format and print data
Usage: printf format [arguments...]
Examples:
  printf "%10s" hello
  printf "%-8.3f" 3.14159
  printf "%x %X %o" 255 255 8
  printf %b "line\\nbreak""
`;

exports[`snapshot: help and type type builtins and external 1`] = `
"printf is a shell builtin
---
env is a shell builtin
---
echo is a shell builtin"
`;
